### PR TITLE
EWPP-26: Extend events filter for ongoing events.

### PR DESCRIPTION
**Pull request by nagyad**

### CODE REVIEW

File: modules/oe_content_event/src/Plugin/InternalLinkSourceFilter/EventPeriodFilter.php [hunk 1]
```diff
@@ -33,6 +33,11 @@
  */
 class EventPeriodFilter extends InternalLinkSourceFilterPluginBase implements InternalLinkSourceFilterInterface, ContainerFactoryPluginInterface {
 
+  /**
+   * Option for ongoing events.
+   */
+  const ONGOING = 2;
```
Comment (sergepavle) [inline modules/oe_content_event/src/Plugin/InternalLinkSourceFilter/EventPeriodFilter.php:39]: Maybe even nicer to use 0 integer value for ongoing event constant (as it is a moment between past and upcoming).

File: modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php [hunk 2]
```diff
@@ -164,6 +176,29 @@ public function testEventLinkSourceFilter(): void {
     ];
     $this->assertEqual($cache->getCacheTags(), $date_cache_tags);
 
+    // Set the time to test ongoing events.
```
Comment (sergepavle) [inline modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php:179]: Looks like we are missing tests of edge cases when the current time is '2016-05-10T12:00:00' or '2050-05-15T12:00:00'. As I understand we are expecting at this moment anyway to have the ongoing event.

File: modules/oe_content_event/src/Plugin/InternalLinkSourceFilter/EventPeriodFilter.php (hunk 1 above)
Comment (nagyad) [reply modules/oe_content_event/src/Plugin/InternalLinkSourceFilter/EventPeriodFilter.php:39]: You can't use 0 there, it will considered as empty and it will show all events.

File: modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php (hunk 2 above)
Comment (nagyad) [reply modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php]: Done.

**Comment by nagyad** [comment]
Rebased.

File: modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php [hunk 3]
```diff
@@ -110,15 +122,16 @@ public function testEventLinkSourceFilter(): void {
     $future_events = [
       $upcoming_event->id() => $upcoming_event->id(),
       $future_event->id() => $future_event->id(),
+      $ongoing_event->id() => $ongoing_event->id(),
     ];
     $this->assertEqual($query_results, $future_events);
```
Comment (imanoleguskiza) [inline modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php:127]: Actually after introducing the ongoing event the two arrays are not equal: the order is different since the ongoing event should be the first.

File: modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php [hunk 4]
```diff
@@ -164,6 +177,31 @@ public function testEventLinkSourceFilter(): void {
...
+    // Set the time to the end date of an ongoing event.
+    $static_time = new DrupalDateTime('2050-05-15 12:00:00', DateTimeItemInterface::STORAGE_TIMEZONE);
+    $time->setTime($static_time->getTimestamp());
+    $query = $storage->getQuery();
+    // Configuring the filter for upcoming events.
+    $plugin->setConfiguration(['period' => EventPeriodFilter::UPCOMING]);
+    $cache = new CacheableMetadata();
+    $plugin->apply($query, [], $cache);
+    $query_results = $query->execute();
+    $ongoing_events = [
```
Comment (imanoleguskiza) [inline modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php:189]: This test makes no sense, according to the comment we "Set the time to the end date of an ongoing event." (which by the way is also the end date for the upcoming event, which makes things even more confusing), so when we use the "upcoming" filter I would assume that the ongoing event SHOULDN'T be in the results, and still it is (together with the upcoming event which has also ended by now).

File: modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php [hunk 5]
```diff
@@ -164,6 +177,31 @@ public function testEventLinkSourceFilter(): void {
...
+      $ongoing_event->id() => $ongoing_event->id(),
+      $upcoming_event->id() => $upcoming_event->id(),
+      $future_event->id() => $future_event->id(),
+    ];
+    $this->assertEqual($query_results, $ongoing_events);
+    // Time based caches have been added and they correspond to the closest
+    // event end date.
+    $date_cache_tags = [
```
Comment (imanoleguskiza) [inline modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php:197]: This also highlights the problem described above, now the cache won't be cleared until 2200 so this means that unless we clear it manually, the ongoing and upcoming events (which have finished) will be there for 150 years :(

File: modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php (hunk 4 above)
Comment (nagyad) [reply modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php:189]: Yeah, but that is the requirement... :(

Comment (nagyad) [reply modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php:189]: We have identified a potential conflict with the fact that we are using the cache tags with the exact end date, but actually the end date is still the date of the event that is ongoing and the granularity of the cache is hour based. Ex.: End date today at 12:00:00 - at this exact time the event is still ongoing the event will become past event when the time hits 12:00:01, but the cache gets invalidated before and that causes inconsistencies. We have to discuss this problem with KUGLER Tamas to understand what should be the minimum granularity of the time caching.

File: modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php (hunk 5 above)
Comment (nagyad) [reply modules/oe_content_event/tests/src/Kernel/EventLinkSourceFilterTest.php:197]: > We have identified a potential conflict with the fact that we are using the cache tags with the exact end date, but actually the end date is still the date of the event that is ongoing and the granularity of the cache is hour based. Ex.: End date today at 12:00:00 - at this exact time the event is still ongoing the event will become past event when the time hits 12:00:01, but the cache gets invalidated before and that causes inconsistencies. We have to discuss this problem with KUGLER Tamas to understand what should be the minimum granularity of the time caching.
